### PR TITLE
Configure testing with Podman and Mamba

### PR DESCRIPTION
**Pull request by TomKellyGenetics**
Implements unit testing and GitHub Actions with Podman. Migrates changes in PR #1706 for separate discussion.

Note that Mamba has also been proposed as an alternative to Conda in PR #2610.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=podman pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=mamba pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: tests/config/nextflow.config [hunk 1]
```diff
@@ -14,12 +14,17 @@ if ("$PROFILE" == "singularity") {
     singularity.enabled = true
     singularity.autoMounts = true
 } else if ("$PROFILE" == "conda") {
-    conda.enabled = true
+    params.enable_conda = true
+    conda.enabled=true
```
Comment (muffato) [inline tests/config/nextflow.config:18]: Restored the initial code as `params.enable_conda` is not used since #2654 (which happened after you started #1706)

```suggestion
    conda.enabled = true
```

**Review by muffato — COMMENTED** [review]
I wouldn't support adding mamba and podman profiles because they both use packages / containers that are already tested through other profiles (resp. conda and docker). In other words, I want the CI to test that the packages / containers defined in each module have a working version of the software (tool), rather than the runtime engine.

This of course assumes that all those runtime engines are interchangeable. I agree with https://github.com/nf-core/modules/pull/1706#discussion_r997075724 : if there are often differences between mamba and conda, or between podman and docker, then yes, we should then probably run both.

mamba is useful because it's faster than conda for installing environments, and that's why I used in #2610. I think there could be a discussion about making the conda profile use mamba for _all_ modules to speed up CI, but adding an extra CI job to test the same packages through mamba would be mostly redundant.

I still left some comments in case others agree with the PR and want to push it

**Comment by jfy133** [comment]
I don't think the `mamba` support is necessary either, given: https://www.anaconda.com/blog/conda-is-fast-now

**Comment by muffato** [comment]
This branch will have to be rebased to include the changes made in #2740

**Comment by TomKellyGenetics** [comment]
Updates as requested. Thanks for taking this proposal into consideration.

File: modules/nf-core/nanoplot/main.nf [hunk 2]
```diff
@@ -2,10 +2,10 @@ process NANOPLOT {
     tag "$meta.id"
     label 'process_low'
 
-    conda "bioconda::nanoplot=1.40.0"
+    conda "bioconda::nanoplot=1.41.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/nanoplot:1.40.0--pyhdfd78af_0' :
-        'quay.io/biocontainers/nanoplot:1.40.0--pyhdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/nanoplot:1.41.0--pyhdfd78af_0' :
+        'quay.io/biocontainers/nanoplot:1.41.0--pyhdfd78af_0' }"
```
Comment (maxulysse) [inline modules/nf-core/nanoplot/main.nf:8]: Can this be in a separate PR?

**Comment by maxulysse** [comment]
Same as other, don't think it's necessary

File: modules/nf-core/nanoplot/main.nf (hunk 2 above)
Comment (muffato) [reply modules/nf-core/nanoplot/main.nf:8]: This change is actually already on master: https://github.com/nf-core/modules/blob/master/modules/nf-core/nanoplot/main.nf#L5-L8
I don't know why GitHub is showing it in this PR. I noticed it's been doing that lately, and it's confusing

File: tests/modules/nf-core/nanoplot/nextflow.config [hunk 3]
```diff
@@ -3,8 +3,3 @@ process {
     publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
 
 }
-
-// Conda takes too long to resolve the dependencies
-conda {
-    useMamba = true
-}
```
Comment (muffato) [inline tests/modules/nf-core/nanoplot/nextflow.config:10]: Similarly, this change is already on master: https://github.com/nf-core/modules/blob/master/tests/modules/nf-core/nanoplot/nextflow.config 
GitHub shouldn't be showing it

File: .github/workflows/pytest-workflow.yml [hunk 4]
```diff
@@ -129,13 +134,17 @@ jobs:
           channels: conda-forge,bioconda,defaults
           python-version: ${{ matrix.python-version }}
 
-      - name: Conda clean
+      - name: Conda setup
```
Comment (muffato) [inline .github/workflows/pytest-workflow.yml:137]: Already on master

File: .github/workflows/pytest-workflow.yml [hunk 5]
```diff
@@ -129,13 +134,17 @@ jobs:
...
         run: conda clean -a
 
       - name: Install mamba
-        if: matrix.profile == 'conda'
+        if: matrix.profile == 'mamba'
         run: conda install mamba -n base -c conda-forge
+        run: |
+          conda clean -a
+          conda install -n base conda-libmamba-solver
+          conda config --set solver libmamba
```
Comment (muffato) [inline .github/workflows/pytest-workflow.yml:147]: Already on master

File: modules/nf-core/nanoplot/main.nf (hunk 2 above)
Comment (TomKellyGenetics) [reply modules/nf-core/nanoplot/main.nf:8]: I pulled the latest version of the remote to "master" branch and confirmed with `git diff master HEAD modules/nf-core/nanoplot/main.nf` and `git checkout master modules/nf-core/nanoplot/main.nf` that there are no changes. These files are identical to those already on master.

File: .github/workflows/pytest-workflow.yml [hunk 6]
```diff
@@ -121,6 +121,11 @@ jobs:
         with:
           singularity-version: 3.7.1
 
+      - name: Set up Podman
+        if: matrix.profile == 'podman'
+        run: |
+          sudo ln -sf /usr/bin/podman /usr/bin/docker
```
[the remark this answers is not in the document]
Comment (TomKellyGenetics) [reply .github/workflows/pytest-workflow.yml:127]: Required by GitHub Actions as discussed here. https://github.com/actions/runner/issues/505#issuecomment-1272719578

**Comment by SPPearce** [comment]
So, do we want to implement Podman and Mamba or do we want to close this PR?
